### PR TITLE
fix(panel): Make all tests wait for promises to resolve.

### PR DESCRIPTION
**Pull request by ErinCoughlan**
1. $mdPanel.open returns a promise that resolves to MdPanelRef. This lets the user chain on open.
2. Tests resolve promises before completing, removing the need for async tests.

@KarenParker @gmoothart @DerekLouie @ThomasBurleson @jelbourn  Please review.

### CODE REVIEW

File: src/components/panel/panel.js [hunk 1]
```diff
@@ -592,7 +593,7 @@ var FOCUS_TRAP_TEMPLATE = angular.element(
 /**
  * A service that is used for controlling/displaying panels on the screen.
  * @param {!angular.JQLite} $rootElement
- * @param {!angular.Scope} $rootScope
+ * @param {!angular.Scope} $rootScope,
```
Comment (KarenParker) [inline src/components/panel/panel.js]: Revert this change.

File: src/components/panel/panel.js [hunk 2]
```diff
@@ -860,7 +860,7 @@ MdPanelRef.prototype.close = function() {
   // TODO(ErinCoughlan) - Cancel any in-progress actions.
 
   var self = this;
-  var closePromise = this.hide()
+  return this.hide()
```
Comment (KarenParker) [inline src/components/panel/panel.js]: Please leave the closePromise, it will be necessary to cancel promises.

File: src/components/panel/panel.js [hunk 3]
```diff
@@ -903,7 +901,7 @@ MdPanelRef.prototype.attachOnly = function() {
 
 
 /**
- * Detaches the panel. Will hide the panel first if visible.
+ * Only detaches the panel. Will NOT hide the panel first.
```
Comment (KarenParker) [inline src/components/panel/panel.js]: Consider also changing the name to detachOnly. Otherwise we should remove the Only from attach.

File: src/components/panel/panel.spec.js [hunk 4]
```diff
@@ -566,15 +565,14 @@ describe('$mdPanel', function() {
       expect(BACKDROP_CLASS).not.toExist();
     });
 
-    it('should show backdrop when hasBackdrop=true', function() {
+    xit('should show backdrop when hasBackdrop=true', function() {
```
Comment (KarenParker) [inline src/components/panel/panel.spec.js]: Please add a TODO(KarenParker)

File: src/components/panel/panel.js (hunk 2 above)
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: It's just a local variable that was immediately returned, making it unnecessary. I reverted the change, but know that WebStorm is unhappy.

File: src/components/panel/panel.js (hunk 3 above)
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: Switched `attachOnly` to `attach` for consistency.

**Comment by ErinCoughlan** [comment]
PTAL. I fixed all the aforementioned comments.

**Comment by KarenParker** [comment]
LGTM

File: src/components/panel/panel.js [hunk 5]
```diff
@@ -1104,16 +1102,19 @@ MdPanelRef.prototype._createPanel = function() {
                 getElement(self._config['attachTo']));
           }
 
-          self._addStyles();
           self._configureTrapFocus();
-          return resolve(self);
+          return self._addStyles().then(function() {
+            self._panelContainer.addClass(MD_PANEL_HIDDEN);
+            return resolve(self);
```
Comment (gmoothart) [inline src/components/panel/panel.js]: Do you still need to return the resolve(self)? That feels like it is mixing 2 styles.

File: src/components/panel/panel.spec.js [hunk 6]
```diff
@@ -1426,46 +1423,45 @@ describe('$mdPanel', function() {
...
+    var openPromise;
     if (opt_config) {
-      panelRef = $mdPanel.open(opt_config);
+      openPromise = $mdPanel.open(opt_config);
     } else if (panelRef) {
-      panelRef.open();
+      openPromise = panelRef.open();
     } else {
-      panelRef = $mdPanel.open();
+      openPromise = $mdPanel.open();
```
Comment (gmoothart) [inline src/components/panel/panel.spec.js]: (nit) This could be just two conditions:
```
if (panelRef) {
  openPromise = panelRef.open();
}
else {
  openPromise = $mdPanel.open(opt_config);
}
```

**Comment by gmoothart** [comment]
:+1: pending comments

File: src/components/panel/panel.js (hunk 5 above)
Comment (ErinCoughlan) [inline src/components/panel/panel.js]: Right, it doesn't need to be returned, just resolved. Nice catch.

File: src/components/panel/panel.js [hunk 7]
```diff
@@ -730,6 +708,25 @@ MdPanelService.prototype.newPanelAnimation = function() {
 };
 
 
+/**
+ * Wraps the users template in two elements, md-panel-container, which covers
```
Comment (ThomasBurleson) [inline src/components/panel/panel.js]: The `div` uses `class="md-panel-outer-wrapper"`. Should this be `md-panel-container` to be consistent with the JS docs ?

Comment (ErinCoughlan) [inline src/components/panel/panel.js]: https://github.com/angular/material/pull/8143
